mm2src/mm2_metrics/src/adapt.rs

```diff
@@ -0,0 +1,381 @@
+            let (name, labels) = key_to_parts(&key, None);
+            let value = histogram.get_inner().data().iter().sum();
+            let entry = histograms
+                .entry(name)
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
+        }
+
+        Snapshot {
+            counters,
+            gauges,
+            histograms,
+        }
+    }
+
+    fn prepare_tag_metrics(&self) -> Vec<PreparedMetric> {
+        let mut output: Vec<PreparedMetric> = vec![];
+
+        for (key, counter) in self.registry.get_counter_handles().drain() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
+        }
+
+        for (key, gauge) in self.registry.get_gauge_handles().drain() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
+        }
+
+        for (key, histogram) in self.registry.get_histogram_handles().drain() {
+            let value: f64 = histogram.get_inner().data().iter().sum();
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value as u64));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
+        }
+
+        output
+    }
+}
+
+#[derive(Clone)]
+pub struct MmRecorder {
+    pub(crate) inner: Arc<Mutex<Inner>>,
+}
+
+impl Default for MmRecorder {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(Inner {
+                registry: Registry::new(metrics_util::registry::GenerationalStorage::atomic()),
+            })),
+        }
+    }
+}
+
+impl From<Inner> for MmRecorder {
+    fn from(inner: Inner) -> Self {
+        MmRecorder {
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+}
+
+impl MmRecorder {
+    pub fn try_recorder(&mut self) -> Option<&mut Self> {
+        if try_recorder().is_some() {
+            return Some(self);
+        };
+        None
+    }
+
+    pub fn handle(&self) -> MmHandle {
```

Comment:
I think this method is not necessary.
I'd delegate the task to either generate Json or print Tags to log from `MmHandle` to `JsonObserver` and `TagsObserver` respectively.
I.e. `JsonObserver::observe(&mm_recorder) -> MetricsJson` and `TagsObserver::observe(&mm_recorder)`.
What do you think?

Reply:
make sense..will remove